appsrc/dcfile/dccp.man
.TH DCCP 1 "22 January 2006" "DICOM PS3" "DICOM PS3 - Copy DICOM file"
.SH NAME
dccp \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Copy DICOM file
.SH SYNOPSIS
.HP 10
.B dccp
[
.B \-v|verbose
]
[
.B \-ignorereaderrors
]
[
.B \-removecommandgroup
]
[
.B \-selectvoilutitem 0..n\-1
]
.so man1/optin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dccp
reads the named dicom or acr-nema input file and copies the information and
pixel data to a new dicom file.
.LP
The input encoding of the file will be automatically determined, or can
be explicitly specified in pathological cases using the input options
described in dcintro(1).
.LP
The encoding of the file will be changed to the default or explicitly
requested transfer syntax, as described in dcintro(1) under output
options. This provides a means of converting between different transfer
syntaxes, and adding or removing Part 10 style meta-information headers
to or from "true" DICOM files, DICOM messages captured from network
exchanges such as with the Mallinckrodt or Oldenburg central test node
software, or converting old ACR/NEMA or SPI files.
.LP
During the translation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows pre-DICOM
ACR-NEMA or SPI files to be converted to DICOM files containing
conforming instances of SOP classes in a semi-automated manner. If the
SOP Class can be automatically determined, such as for CT or MR images,
specific SOP Class instances will be created, otherwise SC will be used.
Unique study, series and instance identifiers will be generated from a
time stamp, but can be overridden by command line options to allow for
shared identifiers between files of a series.
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.TP
.B \-ignorereaderrors
.RS
Normally dccp stops if read errors are encountered, but if this option is
turned on, it will try to write what it can. This may be useful to remove
trailing garbage after a valid dicom set that has been padded past the
pixel data (e.g. to a blocksize) or similar.
.RE
.TP
.B \-removecommandgroup
.RS
Normally dccp will created an invalid copy if the input is in implicit VR and erroneously has
command group (0x0000) elements at the head of the dataset (e.g. if the both the command
and dataset network PDUs were saved to a file), since the command group elements will get
inserted in fromt of the meta information header. Use this option to remove command group
elements (whether or not a meta information header is added).
.RE
.TP
.B \-selectvoilutitem 0..n\-1
.RS
Keep only the selected VOI LUT Sequence Item, where the items are numbered from 0.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% echo Adds meta-information header by default
.RE
% dccp -v NM.dc3 NM2.dc3
.RE
******** While reading ... ********
.RE
@0x00000000,0x00000000 of 0xffffffff: (0x0008,0x0000) ...
.RE
...
.RE
******** As read ... ********
.RE
(0x0008,0x0000) UL Group Length VR=<UL> ...
.RE
...
.RE
******** As written ... ********
.RE
(0x0002,0x0000) UL Meta Element Group Length ...
.RE
(0x0002,0x0001) OB File Meta Information Version ...
.RE
...
.RE
(0x0008,0x0008) CS Image Type VR=<CS> VL=<24> <...>
.RE
...
.RE
\ 
.RE
% echo Make explicit big endian, with stamp & replacement,
.RE
       and removal of private attributes
.RE
\ 
.RE
% dccp \-v \-ts 1.2.840.10008.1.2.2 \-stamp "9999" 
.RE
       \-r PatientSex F \-removeprivate
.RE
       test.dc3 test.big
.RE
******** While reading ... ********
.RE
@0x00000000,0x00000000 of 0xffffffff: (0x0008,0x0000) ...
.RE
...
.RE
******** As read ... ********
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <M > 
.RE
...
.RE
(0x0028,0x0010) US Rows       VR=<US> VL=<2> [100]
.RE
...
.RE
(0x0029,0x0010) LO Private Creator ... [ACME CO ]
.RE
(0x0029,0x1010) DS Zoom       VR=<DS> VL=<0> <>
.RE
...
.RE
******** After replace before... ********
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <F> 
.RE
...
.RE
******** As written ... ********
.RE
(0x0002,0x0000) UL Meta Element Group Length ...
.RE
(0x0002,0x0001) OB File Meta Information Version ...
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <F> 
.RE
...
.RE
(0x0028,0x0010) US Rows       VR=<US> VL=<2> [100]
.RE
...
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dccp.cc
#include "attrmxls.h"
#include "attrseq.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"

void
keepOnlySelectedLUTSequenceItem(AttributeList &list,Tag lutSequenceTag,int whichLUTItemToKeep,TextOutputStream &log) {
	Attribute *aLUTSequence = list[lutSequenceTag];
	if (aLUTSequence) {
		AttributeList **items;
		int nLUTItems=aLUTSequence->getLists(&items);
		if (whichLUTItemToKeep >= 0 && whichLUTItemToKeep < nLUTItems) {
			AttributeList *itemToKeep = items[whichLUTItemToKeep];
			SequenceAttribute *newLUTSequence = new SequenceAttribute(lutSequenceTag);
			Assert(newLUTSequence);
			(*newLUTSequence)+=itemToKeep;
			list-=lutSequenceTag;
			list+=newLUTSequence;
		}
		else {
			log << WMsgDC("Missing LUT Sequence Item") << " - " << lutSequenceTag << " # " << whichLUTItemToKeep << endl;
		}
	}
	else {
		log << WMsgDC("Missing LUT Sequence") << " - " << lutSequenceTag << endl;
	}
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool ignorereaderrors=options.get("ignorereaderrors");
	bool removecommandgroup=options.get("removecommandgroup");

	int whichVOILUTitem=0;
	bool selectVOILUTitem=options.get("selectvoilutitem",whichVOILUTitem);

	dicom_input_options.done();
	dicom_output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< dicom_output_options.usage()
			<< " [-v|-verbose]"
			<< " [-ignorereaderrors]"
			<< " [-removecommandgroup]"
			<< " [-selectvoilutitem 0..n-1]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream  log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	if (!list.good()) {
		if (!ignorereaderrors) {
			log << list.errors();
			success=false;
		}
		log << EMsgDC(DatasetReadFailed) << endl;
	}

	if (success) {
		// Don't write list warnings yet ...
		// done in usualManagedAttributeListWrite ...
		
		if (removecommandgroup) {
			list.removeCommandGroup();
		}
		
		if (selectVOILUTitem) {
			keepOnlySelectedLUTSequenceItem(list,TagFromName(VOILUTSequence),whichVOILUTitem,log);
		}

		if (!usualManagedAttributeListWrite(list,dout,
			dicom_output_options,log,verbose)) success=false;
	}

	return success ? 0 : 1;
}
